### ISSUE TITLE
JIT: Using @Optional() private causes wrong Renderer to be injected 

### ISSUE THREAD

**Issue by andrewseguin** [angular/angular#17378]
In Angular Material we are seeing an odd issue where the injection system is using the old Renderer when it isn't supposed to.

Our MdChips component uses Renderer2 but it receives Renderer1 if the parent component also has an MdInputContainer with an input containing an ngModel.

To reproduce, run `gulp test` on Angular Material using the branch found here: https://github.com/andrewseguin/material2/tree/injection-issue

The changes that reproduced the error can be found in this commit: https://github.com/andrewseguin/material2/commit/d9cd4c642a7ec698e1c1cb5ff0953bc4adc0a460

**Comment by jelbourn** [comment]
To clarify, the simplest reproduction looks like (captured in the reproduction):
```html
<md-chip-list>
  <md-chip *ngFor="let item of list>...</md-chip>
</md-chip-list>

<md-input-container>
  <input mdInput [(ngModel)]="someValue">
</md-input-container>
```
Where the constructor for `MdInput` has
```ts
  constructor(private _elementRef: ElementRef,
              private _renderer: Renderer2,
              @Optional() @Self() public _ngControl: NgControl,
              @Optional() private _parentForm: NgForm,
              @Optional() private _parentFormGroup: FormGroupDirective) {
```

The problem **only** happens when:
* The input has as `NgModel` applied
* The chip is on the same element as `NgFor`
* The `list` iterated over by `NgFor` changes after `OnInit`
* `MdInput` injects `FormGroupDirective` with `@Optional` **and** `private`


My best guess is that having a decorator on a private member somehow triggers a code path that corrupts the injector for the view (since the chips are siblings/cousins to the input)

**Comment by edcarroll** [comment]
I've started running into this issue in my library as well, however in this instance, I'm not using any decorators in the component (which attaches to `.item`) constructor:

```typescript
constructor(private _renderer:Renderer2, public element:ElementRef) {
```

The simplest reproduction we've managed in [this plunkr](https://plnkr.co/edit/tnhxg75ibVORuidb4xBw).

Unfortunately it's incredibly vague as to when it happens and doesn't, and so far I can't seem to get a definitive list of requirements for reproduction.

**Comment by bgotink** [comment]
I've got the same issue, though my reproduction ([plunker link](https://plnkr.co/edit/z5SGnEDdd6NhpFVG1GZA)) is slightly different

```html
<form myForm>
  <my-form-group>
    <my-form-element>
      <input type="checkbox" [(ngModel)]="checked" name="foo">
      <my-validation *ngIf="!checked">foo</my-validation>
    </my-form-element>
  </my-form-group>
</form>
```

with the `MyValidation` constructor

```typescript
constructor(renderer: Renderer2, elementRef: ElementRef) {
  console.log('validation:', renderer);
}
```

This results in

![image](https://user-images.githubusercontent.com/821510/27377497-6c2a4a06-5675-11e7-9647-56ac35143b18.png)

It took me a while to get this reproduction running, because there are a lot of conditions before this issue arises:

- the `[myForm]` component must be present on the form element
- the `*ngIf` must be present on the validation component
- the `[(ngModel)]` must be present on the input
- the `ElementRef` must be injected in the `MyValidationComponent` constructor
- the `providers` mustn't be empty in the `MyFormElementComponent`'s `@Component` metadata
- the `<my-form-group>` element must be in the template

Change any one of these conditions and suddenly the correct `Renderer2` is injected.

![image](https://user-images.githubusercontent.com/821510/27378000-cfbeeb5c-5676-11e7-862d-379caae43a48.png)

**Comment by tytskyi** [comment]
Hi @andrewseguin i checked your reproduction - this looks fixed with angular `4.3.1`. Test is green

**Comment by andrewseguin** [comment]
Awesome thanks @tytskyi
